File: IT/2/Interfacce grafiche/Esercizi/Java/Esercizio30.java
/* file Esercizio30.java */

import java.awt.Dimension;
import java.awt.GridBagLayout;
import java.awt.GridBagConstraints;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JPanel;
import javax.swing.JTextPane;
import javax.swing.text.Style;
import javax.swing.text.StyleContext;
import javax.swing.text.StyledDocument;
import javax.swing.text.DefaultStyledDocument;
import javax.swing.text.StyleConstants;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.Stack;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class Esercizio30 {
	private static class KPad extends JPanel implements ActionListener {
		private Calc calc;
		private String [] btns = { "7", "8", "9", "\u00F7", "CE", "C", "4", "5", "6", "\u2217", "(", ")", "1", "2", "3", "-", "x\u00b2", "\u221A", "0", ".", "%", "+", "=" };

		public KPad(JFrame parent) {
			super(new GridBagLayout());
			this.calc = (Calc) parent;
			int r = 0;
			int c = 0;
			for (String btn : btns) {
				JButton b = new JButton(btn);
				b.addActionListener(this);
				GridBagConstraints gbc = new GridBagConstraints();
				gbc.gridx = c;
				gbc.gridy = r;
				gbc.fill = GridBagConstraints.BOTH;
				if (!btn.equals("=")) c++;
				else {
					gbc.gridwidth = 2;
					c += 2;
				}
				add(b, gbc);

				if (c % 6 == 0) {
					c = 0;
					r++;
				}
			}
		}

		@Override
		public void actionPerformed(ActionEvent e) {
			String s = ((JButton) e.getSource()).getText();
			StyledDocument sd = calc.scr.getStyledDocument();
			if (s.equals("=")) {
				try {
					String expr = calc.scr.getText();
					List<String> rpn = calc.infixToRpn(expr);
					Double result = calc.evaluate(rpn);
					this.calc.scr.setText(String.valueOf(result));
				}
				catch (Exception ex) {
					calc.scr.setText("Espressione malformata");
				}
			}
			else if (s.equals("C")) {
				this.calc.scr.setText("");
			}
			else if (s.equals("CE")) {
				String onsc = calc.scr.getText();
				String lastN = "";
				Matcher m = Pattern.compile("[\\d]+(\\.[\\d]+)?").matcher(onsc);
				while (m.find()) lastN = m.group();
				int i = onsc.lastIndexOf(lastN);
				this.calc.scr.setText(onsc.substring(0, i));
			}
			else {
				// quadrato
				if (s.equals("x\u00b2")) s = "\u00b2";
				try {
					Style style = sd.getStyle(StyleContext.DEFAULT_STYLE);
					sd.insertString(sd.getLength(), s, style);
				}
				catch (Exception ex) {
					System.err.println(ex);
				}
			}
		}
	}

	private static class Calc extends JFrame {
		// mappa degli operatori con precedenza e associativà
		private static Map<String, int[]> ops = new HashMap<String, int[]>() {
			{
				put("+", new int[] {0, 0});
				put("-", new int[] {0, 0});
				put("/", new int[] {1, 0});
				put("*", new int[] {1, 0});
				put("^", new int[] {2, 1});
			}
		};

		public JTextPane scr;

		public Calc() {
			getContentPane().setLayout(new BoxLayout(getContentPane(), BoxLayout.Y_AXIS));
			StyleContext ctx = new StyleContext();
			StyledDocument doc = new DefaultStyledDocument(ctx);
			Style style = ctx.getStyle(ctx.DEFAULT_STYLE);
			StyleConstants.setAlignment(style, StyleConstants.ALIGN_RIGHT);
			this.scr = new JTextPane(doc);
			add(scr);
			add(new KPad(this));
			pack();
			//setSize(new Dimension(300, 300));
			setSize(new Dimension(280, 225));
			setResizable(false);
			setTitle("Calcolatrice");
			setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
			setVisible(true);
		}

		public static List<String> infixToRpn(String expr) throws Exception {
			expr = expr.replace("\u00b2", "^2").replace("\u221A", "sqrt").replace("\u00F7", "/").replace("\u2217", "*");
			Matcher m = Pattern.compile("^[-+]?[\\d]+(\\.[\\d]+)?%?|[\\d]+(\\.[\\d]+)?%?|sqrt[\\d]+(\\.[\\d]+)?|[()*/+-\\^]").matcher(expr);
			List<String> tokens = new ArrayList<>();

			while (m.find()) {
				tokens.add(m.group());
			}

			List<String> rpn = new ArrayList<>();
			Stack<String> stackOps = new Stack<>();

			for (String token : tokens) {
				if (ops.containsKey(token)) {
					int [] prop = ops.get(token);
					int prec = prop[0];
					int ass = prop[1];

					while (!stackOps.empty() && ops.containsKey(stackOps.peek())) {
						int opPrec = ops.get(stackOps.peek())[0];
						if (prec < opPrec || (prec == opPrec && ass == 0)) {
							rpn.add(stackOps.pop());
						}
						break;
					}
					stackOps.push(token);
				}
				else if (token.equals("(")) {
					stackOps.push(token);
				}
				else if (token.equals(")")) {
					while (!stackOps.empty() && !stackOps.peek().equals("(")) {
						rpn.add(stackOps.pop());
					}
					stackOps.pop();
				}
				else { // numero o funzione
					// se si tratta di una funzione, la valuta e salva subito il risultato
					if (token.indexOf("sqrt") == 0) {
						Matcher f = Pattern.compile("[\\d]+(\\.[\\d]+)?").matcher(token);
						if (f.find()) {
							String s = f.group();
							Double d = Double.valueOf(s);
							Double res = Math.sqrt(d);
							token = String.valueOf(res);
						}
					}
					rpn.add(token);
				}
			}

			while (!stackOps.empty()) {
				rpn.add(stackOps.pop());
			}

			return rpn;
		}

		public static Double evaluate(List<String> rpn) throws Exception {
			Stack<String> stack = new Stack<>();
			for (String token : rpn) {
				if (!ops.containsKey(token)) {
					stack.push(token);
				}
				else {
					String s = stack.pop();
					Double o2 = 0.0;
					Double o1 = Double.valueOf(stack.pop());

					if (!s.endsWith("%")) o2 = Double.valueOf(s);
					else {
						// calcola la percentuale
						s = s.replace("%", "");
						o2 = o1/100*Double.valueOf(s);
					}

					Double result = 0.0;
					if (token.equals("+")) result = o1 + o2;
					else if (token.equals("-")) result = o1 - o2;
					else if (token.equals("*")) result = o1 * o2;
					else if (token.equals("/")) result = o1 / o2;
					else if (token.equals("^")) result = Math.pow(o1, o2);

					stack.push(String.valueOf(result));
				}
			}

			return Double.valueOf(stack.pop());
		}
	}

	public static void main(String [] args) {
		Calc calc = new Calc();
	}
}

File: IT/2/Interfacce grafiche/Esercizi/Python/esercizio30.py
# file esercizio30.py

from __future__ import unicode_literals, division

try:
	from tkinter import Frame, Text, Button, END
except ImportError:
	from Tkinter import Frame, Text, Button, END

import re
import math

class MyPanel(Frame):
	def __init__(self, parent): 
		self.buttons = ["7", "8", "9", u"\u00F7", "CE", "C", "4", "5", "6", u"\u2217", "(", ")", "1", "2", "3", "-", u"x\u00b2", u"\u221A", "0", ".", "%", "+", "="]
		Frame.__init__(self, master=parent, width=300, height=300)

		r = 1
		c = 0
		for s in self.buttons:
			b = Button(self, text=s)
			b.bind("<Button-1>", self.clickHandler)
			if s != "=":
				b.grid(row=r, column=c, sticky="ew")
				c += 1
			else:
				b.grid(row=r, column=c, columnspan=2, sticky="ew")
				c += 2
			if c % 6 == 0:
				c = 0
				r += 1

		self.pack()

	def clickHandler(self, event):
		s = event.widget["text"]
		if s == "=":
			try:
				expr = self.master.s.get(1.0, END).rstrip()
				rpn = self.master.infixToRpn(expr)
				result = self.master.evaluate(rpn)
				self.master.s.delete(1.0, END)
				self.master.s.insert(1.0, result)
				self.master.s.tag_add("right", 1.0, END)
			except Exception:
				# gestione espressioni malformate
				self.master.s.delete(1.0, END)
				self.master.s.insert(1.0, "Espressione malformata")
				self.master.s.tag_add("right", 1.0, END)
		elif s == "C":
			self.master.s.delete(1.0, END)
		elif s == "CE":
			onsc = self.master.s.get(1.0, END).rstrip()
			n = []
			for m in re.finditer(r"[\d]+(\.[\d]+)?", onsc):
				n.append(m.group())
			last = n[-1]
			i = onsc.rfind(last)
			self.master.s.delete(1.0, END)
			self.master.s.insert(1.0, onsc[:i])
			self.master.s.tag_add("right", 1.0, END)
		else:
			if self.master.s.get(1.0, END).rstrip() == "Espressione malformata":
				self.master.s.delete(1.0, END)

			# elevamento al quadrato
			elif s == "x\u00b2": s = "\u00b2"
			self.master.s.insert(END, s)
			self.master.s.tag_add("right", 1.0, END)

class Esercizio30(Frame):
	def __init__(self):
		Frame.__init__(self, width=300, height=300)

		self.ops = {"+" : [0, 0], "-" : [0, 0], "*" : [1, 0], "/" : [1, 0], "^" : [2, 1]}
		self.s = Text(self, width=30, height=5)
		self.s.tag_configure("right", justify="right")
		self.s.pack()
		self.p = MyPanel(self)
		self.pack()
		self.master.wm_title("Calcolatrice")
		self.mainloop()

	def infixToRpn(self, expr):
		expr = expr.replace("\u00b2", "^2").replace("\u221A", "sqrt").replace("\u00F7", "/").replace("\u2217", "*")
		reg = r"^[-+]?[\d]+(\.[\d]+)?%?|[\d]+(\.[\d]+)?%?|sqrt[\d]+(\.[\d]+)?|[()*/+-\^]"
		tokens = []
		for m in re.finditer(reg, expr):
			tokens.append(m.group())

		rpn = []
		stackOps = []

		for token in tokens:
			if token in self.ops.keys():
				prop = self.ops[token]
				prec = prop[0]
				ass = prop[1]

				while stackOps and stackOps[-1] in self.ops.keys():
					opPrec = self.ops[stackOps[-1]][0]
					if prec < opPrec or prec == opPrec and ass == 0:
						rpn.append(stackOps.pop())
					break
				stackOps.append(token)
			elif token == "(":
				stackOps.append(token)
			elif token == ")":
				while stackOps and stackOps[-1] != "(":
					rpn.append(stackOps.pop())
				stackOps.pop()
			else: # numero o funzione
				# se si tratta di una funzione, la valuta e salva subito il risultato
				if token.find("sqrt") == 0:
					for m in re.finditer(r"[\d]+(\.[\d]+)?", token):
						f = m.group()
					if f:
						d = float(f)
						res = math.sqrt(d)
						token = str(res)
				rpn.append(token)

		while stackOps:
			rpn.append(stackOps.pop())

		return rpn

	def evaluate(self, rpn):
		stack = []
		for token in rpn:
			if token not in self.ops.keys():
				stack.append(token)
			else:
				s = stack.pop()
				o2 = 0.0
				o1 = float(stack.pop())

				if not s[-1] == "%": o2 = float(s)
				else:
					# calcola la percentuale
					s = s.replace("%", "")
					o2 = o1/100*float(s)

				result = 0.0
				if token == "+": result = o1 + o2
				elif token == "-": result = o1 - o2
				elif token == "*": result = o1 * o2
				elif token == "/": result = o1 / o2
				elif token == "^": result = o1**o2

				stack.append(result)

		return float(stack.pop())

calc = Esercizio30()

